File: gateOne/daySeven/bankeBank/BankeApp.java
package daySeven.bankeBank;

import java.util.HashMap;
import java.util.Map;

public class BankeApp {
    private final Map<String, Account> accounts = new HashMap<>();
    private Account activeAccount;

    public String createAccount(String firstName, String lastName, int pin) {
        validateName(firstName);
        validateName(lastName);
        validatePin(pin);
        String accountNumber = generateAccountNumber();
        Account newAccount = new Account(firstName, lastName, pin, accountNumber);
        accounts.put(accountNumber, newAccount);
        activeAccount = newAccount;
        return accountNumber;
    }

    public void login(String accountNumber, int pin) {
        Account account = accounts.get(accountNumber);
        if (account == null) {
            throw new IllegalArgumentException("Account not found");
        }
        if (!account.verifyPin(pin)) {
            throw new IllegalArgumentException("Incorrect PIN");
        }
        activeAccount = account;
    }

    public void logout() {
        activeAccount = null;
    }

    public void deposit(double amount) {
        checkActiveAccount();
        activeAccount.deposit(amount);
    }

    public void withdraw(double amount, int pin) {
        checkActiveAccount();
        activeAccount.withdraw(amount, pin);
    }

    public double getBalance() {
        checkActiveAccount();
        return activeAccount.getBalance();
    }

    public void transfer(String recipientAccountNumber, double amount, int pin) {
        checkActiveAccount();
        Account recipient = accounts.get(recipientAccountNumber);
        if (recipient == null) {
            throw new IllegalArgumentException("Recipient account does not exist");
        }
        activeAccount.withdraw(amount, pin);
        recipient.deposit(amount);
    }

    public void changePin(int oldPin, int newPin) {
        checkActiveAccount();
        activeAccount.changePin(oldPin, newPin);
    }

    public boolean verifyPin(int pin) {
        checkActiveAccount();
        return activeAccount.verifyPin(pin);
    }

    public void closeAccount(int pin) {
        checkActiveAccount();
        if (!activeAccount.verifyPin(pin)) {
            throw new IllegalArgumentException("Incorrect PIN");
        }
        accounts.remove(activeAccount.getAccountNumber());
        activeAccount = null;
    }

    public String getAccountName() {
        checkActiveAccount();
        return activeAccount.getFirstName() + " " + activeAccount.getLastName();
    }

    public String getAccountNumber() {
        checkActiveAccount();
        return activeAccount.getAccountNumber();
    }

    public int getAccountCount() {
        return accounts.size();
    }

    private String generateAccountNumber() {
        return String.valueOf(1000000000 + (int) (Math.random() * 900000000));
    }

    private void validateName(String name) {
        if (!name.matches("^[A-Za-z]+$")) {
            throw new IllegalArgumentException("Name should contain only letters");
        }
    }

    private void validatePin(int pin) {
        if (String.valueOf(pin).length() != 4 || !String.valueOf(pin).matches("^[0-9]+$")) {
            throw new IllegalArgumentException("PIN must be exactly 4 digits");
        }
    }

    private void checkActiveAccount() {
        if (activeAccount == null) {
            throw new IllegalStateException("No active account");
        }
    }
}

File: gateOne/daySeven/bankeBank/Python/BankeApp.py
import random
import re
from Account import Account

class BankeApp:
    def __init__(self):
        self._accounts = {}
        self._active_account = None

    def create_account(self, first_name, last_name, pin):
        self._validate_name(first_name)
        self._validate_name(last_name)
        self._validate_pin(pin)
        account_number = self._generate_account_number()
        new_account = Account(first_name, last_name, pin, account_number)
        self._accounts[account_number] = new_account
        self._active_account = new_account
        return account_number

    def login(self, account_number, pin):
        account = self._accounts.get(account_number)
        if account is None:
            raise ValueError("Account not found")
        if not account.verify_pin(pin):
            raise ValueError("Incorrect PIN")
        self._active_account = account

    def logout(self):
        self._active_account = None

    def deposit(self, amount):
        self._check_active_account()
        self._active_account.deposit(amount)

    def withdraw(self, amount, pin):
        self._check_active_account()
        self._active_account.withdraw(amount, pin)

    def get_balance(self):
        self._check_active_account()
        return self._active_account.get_balance()

    def transfer(self, recipient_account_number, amount, pin):
        self._check_active_account()
        recipient = self._accounts.get(recipient_account_number)
        if recipient is None:
            raise ValueError("Recipient account does not exist")
        if recipient_account_number == self._active_account.get_account_number():
            raise ValueError("Cannot transfer to the same account")
        self._active_account.withdraw(amount, pin)
        recipient.deposit(amount)

    def change_pin(self, old_pin, new_pin):
        self._check_active_account()
        self._active_account.change_pin(old_pin, new_pin)

    def verify_pin(self, pin):
        self._check_active_account()
        return self._active_account.verify_pin(pin)

    def close_account(self, pin):
        self._check_active_account()
        if not self._active_account.verify_pin(pin):
            raise ValueError("Incorrect PIN")
        del self._accounts[self._active_account.get_account_number()]
        self._active_account = None

    def get_account_name(self):
        self._check_active_account()
        return f"{self._active_account.get_first_name()} {self._active_account.get_last_name()}"

    def get_account_number(self):
        self._check_active_account()
        return self._active_account.get_account_number()

    def get_account_count(self):
        return len(self._accounts)

    def _generate_account_number(self):
        return str(1000000000 + random.randint(0, 900000000))

    def _validate_name(self, name):
        if not re.match("^[A-Za-z]+$", name):
            raise ValueError("Name should contain only letters")

    def _validate_pin(self, pin):
        if len(str(pin)) != 4 or not str(pin).isdigit():
            raise ValueError("PIN must be exactly 4 digits")

    def _check_active_account(self):
        if self._active_account is None:
            raise ValueError("No active account")